// @progress/kendo-angular-common/dist/npm/draggable/draggable.module.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var draggable_directive_1 = require("./draggable.directive");
var common_1 = require("@angular/common");
/**
 * @hidden
 */
var ɵngcc0 = require('@angular/core');
var ɵngcc1 = require('./draggable.directive');
var ɵngcc2 = require('@angular/common');
var DraggableModule = /** @class */ (function () {
    function DraggableModule() {
    }
DraggableModule.ngModuleDef = ɵngcc0.ɵɵdefineNgModule({ type: DraggableModule });
/*@__PURE__*/ ɵngcc0.ɵɵsetNgModuleScope(DraggableModule, { declarations: [ɵngcc1.DraggableDirective], imports: [ɵngcc2.CommonModule], exports: [ɵngcc1.DraggableDirective] });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(DraggableModule, [{
        type: NgModule,
        args: [{
                declarations: [draggable_directive_1.DraggableDirective],
                exports: [draggable_directive_1.DraggableDirective],
                imports: [common_1.CommonModule]
            }]
    }], function () { return []; }, null);
DraggableModule.ngInjectorDef = ɵngcc0.ɵɵdefineInjector({ factory: function DraggableModule_Factory(t) { return new (t || DraggableModule)(); }, imports: [[common_1.CommonModule]] });
    return DraggableModule;
}());
exports.DraggableModule = DraggableModule;
