// @progress/kendo-angular-inputs/dist/es/textbox/localization/custom-messages.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { TextBoxMessages } from './messages';
/**
 * Custom component messages override default component messages.
 */
import * as ɵngcc0 from '@angular/core';
var TextBoxCustomMessagesComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TextBoxCustomMessagesComponent, _super);
    function TextBoxCustomMessagesComponent(service) {
        var _this = _super.call(this) || this;
        _this.service = service;
        return _this;
    }
    TextBoxCustomMessagesComponent_1 = TextBoxCustomMessagesComponent;
    Object.defineProperty(TextBoxCustomMessagesComponent.prototype, "override", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    var TextBoxCustomMessagesComponent_1;
    TextBoxCustomMessagesComponent = TextBoxCustomMessagesComponent_1 = tslib_1.__decorate([
        tslib_1.__metadata("design:paramtypes", [LocalizationService])
    ], TextBoxCustomMessagesComponent);
TextBoxCustomMessagesComponent.ngComponentDef = ɵngcc0.ɵɵdefineComponent({ type: TextBoxCustomMessagesComponent, selectors: [["kendo-textbox-messages"]], factory: function TextBoxCustomMessagesComponent_Factory(t) { return new (t || TextBoxCustomMessagesComponent)(ɵngcc0.ɵɵdirectiveInject(LocalizationService)); }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: TextBoxMessages,
                useExisting: forwardRef(function () { return TextBoxCustomMessagesComponent_1; }) // tslint:disable-line:no-forward-ref
            }
        ]), ɵngcc0.ɵɵInheritDefinitionFeature], consts: 0, vars: 0, template: function TextBoxCustomMessagesComponent_Template(rf, ctx) { }, encapsulation: 2 });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(TextBoxCustomMessagesComponent, [{
        type: Component,
        args: [{
                providers: [
                    {
                        provide: TextBoxMessages,
                        useExisting: forwardRef(function () { return TextBoxCustomMessagesComponent_1; }) // tslint:disable-line:no-forward-ref
                    }
                ],
                selector: 'kendo-textbox-messages',
                template: ""
            }]
    }], function () { return [{ type: LocalizationService }]; }, { override: [] });
    return TextBoxCustomMessagesComponent;
}(TextBoxMessages));
export { TextBoxCustomMessagesComponent };
