// @progress/kendo-angular-intl/dist/es2015/intl.module.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { DatePipe } from './date.pipe';
import { NumberPipe } from './number.pipe';
import * as ɵngcc0 from '@angular/core';
const pipes = [
    DatePipe,
    NumberPipe
];
/**
 * Represents the [NgModule]({{ site.data.urls.angular['ngmoduleapi'] }}docs/ts/latest/guide/ngmodule.html)
 * definition for the Intl services.
 */
let IntlModule = class IntlModule {
}
IntlModule.ngModuleDef = ɵngcc0.ɵɵdefineNgModule({ type: IntlModule });
/*@__PURE__*/ ɵngcc0.ɵɵsetNgModuleScope(IntlModule, { declarations: [DatePipe,
        NumberPipe], exports: [DatePipe,
        NumberPipe] });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(IntlModule, [{
        type: NgModule,
        args: [{
                declarations: [pipes],
                exports: [pipes]
            }]
    }], null, null);
IntlModule.ngInjectorDef = ɵngcc0.ɵɵdefineInjector({ factory: function IntlModule_Factory(t) { return new (t || IntlModule)(); } });;
export { IntlModule };
