// @progress/kendo-angular-inputs/dist/es/slider/localization/custom-messages.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { SliderMessages } from './messages';
/**
 * Custom component messages override default component messages.
 */
import * as ɵngcc0 from '@angular/core';
var SliderCustomMessagesComponent = /** @class */ (function (_super) {
    tslib_1.__extends(SliderCustomMessagesComponent, _super);
    function SliderCustomMessagesComponent(service) {
        var _this = _super.call(this) || this;
        _this.service = service;
        return _this;
    }
    SliderCustomMessagesComponent_1 = SliderCustomMessagesComponent;
    Object.defineProperty(SliderCustomMessagesComponent.prototype, "override", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    var SliderCustomMessagesComponent_1;
    SliderCustomMessagesComponent = SliderCustomMessagesComponent_1 = tslib_1.__decorate([
        tslib_1.__metadata("design:paramtypes", [LocalizationService])
    ], SliderCustomMessagesComponent);
SliderCustomMessagesComponent.ngComponentDef = ɵngcc0.ɵɵdefineComponent({ type: SliderCustomMessagesComponent, selectors: [["kendo-slider-messages"]], factory: function SliderCustomMessagesComponent_Factory(t) { return new (t || SliderCustomMessagesComponent)(ɵngcc0.ɵɵdirectiveInject(LocalizationService)); }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: SliderMessages,
                useExisting: forwardRef(function () { return SliderCustomMessagesComponent_1; }) // tslint:disable-line:no-forward-ref
            }
        ]), ɵngcc0.ɵɵInheritDefinitionFeature], consts: 0, vars: 0, template: function SliderCustomMessagesComponent_Template(rf, ctx) { }, encapsulation: 2 });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(SliderCustomMessagesComponent, [{
        type: Component,
        args: [{
                providers: [
                    {
                        provide: SliderMessages,
                        useExisting: forwardRef(function () { return SliderCustomMessagesComponent_1; }) // tslint:disable-line:no-forward-ref
                    }
                ],
                selector: 'kendo-slider-messages',
                template: ""
            }]
    }], function () { return [{ type: LocalizationService }]; }, { override: [] });
    return SliderCustomMessagesComponent;
}(SliderMessages));
export { SliderCustomMessagesComponent };
