// @progress/kendo-angular-common/dist/npm/events/events.module.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var events_outside_angular_directive_1 = require("./events-outside-angular.directive");
/**
 * @hidden
 */
var ɵngcc0 = require('@angular/core');
var ɵngcc1 = require('./events-outside-angular.directive');
var EventsModule = /** @class */ (function () {
    function EventsModule() {
    }
EventsModule.ngModuleDef = ɵngcc0.ɵɵdefineNgModule({ type: EventsModule });
/*@__PURE__*/ ɵngcc0.ɵɵsetNgModuleScope(EventsModule, { declarations: [ɵngcc1.EventsOutsideAngularDirective], exports: [ɵngcc1.EventsOutsideAngularDirective] });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(EventsModule, [{
        type: NgModule,
        args: [{
                declarations: [events_outside_angular_directive_1.EventsOutsideAngularDirective],
                exports: [events_outside_angular_directive_1.EventsOutsideAngularDirective]
            }]
    }], function () { return []; }, null);
EventsModule.ngInjectorDef = ɵngcc0.ɵɵdefineInjector({ factory: function EventsModule_Factory(t) { return new (t || EventsModule)(); } });
    return EventsModule;
}());
exports.EventsModule = EventsModule;
