// @progress/kendo-angular-inputs/dist/es/rangeslider/localization/custom-messages.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { RangeSliderMessages } from './messages';
/**
 * Custom component messages override default component messages.
 */
import * as ɵngcc0 from '@angular/core';
var RangeSliderCustomMessagesComponent = /** @class */ (function (_super) {
    tslib_1.__extends(RangeSliderCustomMessagesComponent, _super);
    function RangeSliderCustomMessagesComponent(service) {
        var _this = _super.call(this) || this;
        _this.service = service;
        return _this;
    }
    RangeSliderCustomMessagesComponent_1 = RangeSliderCustomMessagesComponent;
    Object.defineProperty(RangeSliderCustomMessagesComponent.prototype, "override", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    var RangeSliderCustomMessagesComponent_1;
    RangeSliderCustomMessagesComponent = RangeSliderCustomMessagesComponent_1 = tslib_1.__decorate([
        tslib_1.__metadata("design:paramtypes", [LocalizationService])
    ], RangeSliderCustomMessagesComponent);
RangeSliderCustomMessagesComponent.ngComponentDef = ɵngcc0.ɵɵdefineComponent({ type: RangeSliderCustomMessagesComponent, selectors: [["kendo-rangeslider-messages"]], factory: function RangeSliderCustomMessagesComponent_Factory(t) { return new (t || RangeSliderCustomMessagesComponent)(ɵngcc0.ɵɵdirectiveInject(LocalizationService)); }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: RangeSliderMessages,
                useExisting: forwardRef(function () { return RangeSliderCustomMessagesComponent_1; }) // tslint:disable-line:no-forward-ref
            }
        ]), ɵngcc0.ɵɵInheritDefinitionFeature], consts: 0, vars: 0, template: function RangeSliderCustomMessagesComponent_Template(rf, ctx) { }, encapsulation: 2 });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(RangeSliderCustomMessagesComponent, [{
        type: Component,
        args: [{
                providers: [
                    {
                        provide: RangeSliderMessages,
                        useExisting: forwardRef(function () { return RangeSliderCustomMessagesComponent_1; }) // tslint:disable-line:no-forward-ref
                    }
                ],
                selector: 'kendo-rangeslider-messages',
                template: ""
            }]
    }], function () { return [{ type: LocalizationService }]; }, { override: [] });
    return RangeSliderCustomMessagesComponent;
}(RangeSliderMessages));
export { RangeSliderCustomMessagesComponent };
