// @progress/kendo-angular-inputs/dist/es/rangeslider/localization/localized-rangeslider-messages.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Directive, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { RangeSliderMessages } from './messages';
/**
 * @hidden
 */
import * as ɵngcc0 from '@angular/core';
var LocalizedRangeSliderMessagesDirective = /** @class */ (function (_super) {
    tslib_1.__extends(LocalizedRangeSliderMessagesDirective, _super);
    function LocalizedRangeSliderMessagesDirective(service) {
        var _this = _super.call(this) || this;
        _this.service = service;
        return _this;
    }
    LocalizedRangeSliderMessagesDirective_1 = LocalizedRangeSliderMessagesDirective;
    var LocalizedRangeSliderMessagesDirective_1;
    LocalizedRangeSliderMessagesDirective = LocalizedRangeSliderMessagesDirective_1 = tslib_1.__decorate([
        tslib_1.__metadata("design:paramtypes", [LocalizationService])
    ], LocalizedRangeSliderMessagesDirective);
LocalizedRangeSliderMessagesDirective.ngDirectiveDef = ɵngcc0.ɵɵdefineDirective({ type: LocalizedRangeSliderMessagesDirective, selectors: [["", "kendoSliderLocalizedMessages", ""]], factory: function LocalizedRangeSliderMessagesDirective_Factory(t) { return new (t || LocalizedRangeSliderMessagesDirective)(ɵngcc0.ɵɵdirectiveInject(LocalizationService)); }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: RangeSliderMessages,
                useExisting: forwardRef(function () { return LocalizedRangeSliderMessagesDirective_1; }) // tslint:disable-line:no-forward-ref
            }
        ]), ɵngcc0.ɵɵInheritDefinitionFeature] });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(LocalizedRangeSliderMessagesDirective, [{
        type: Directive,
        args: [{
                providers: [
                    {
                        provide: RangeSliderMessages,
                        useExisting: forwardRef(function () { return LocalizedRangeSliderMessagesDirective_1; }) // tslint:disable-line:no-forward-ref
                    }
                ],
                selector: '[kendoSliderLocalizedMessages]'
            }]
    }], function () { return [{ type: LocalizationService }]; }, null);
    return LocalizedRangeSliderMessagesDirective;
}(RangeSliderMessages));
export { LocalizedRangeSliderMessagesDirective };
