// @progress/kendo-angular-inputs/dist/es/shared/input-separator.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Component, HostBinding } from '@angular/core';
/**
 * Specifies a separator in the content of components like the TextArea and the TextBox. ([see examples]({% slug adornments_textbox %}#toc-separator)).
 * @example
 * ```ts-no-run
 * _@Component({
 * selector: 'my-app',
 * template: `
 *  <kendo-textbox>
 *    <ng-template kendoTextBoxSuffixTemplate>
 *        <kendo-input-separator></kendo-input-separator>
 *        <button kendoButton look="clear" icon="image"></button>
 *    </ng-template>
 *  </kendo-textbox>
 * `
 * })
 * class AppComponent {}
 * ```
 */
import * as ɵngcc0 from '@angular/core';
var InputSeparatorComponent = /** @class */ (function () {
    function InputSeparatorComponent() {
        this.hostClass = true;
    }
    tslib_1.__decorate([
        HostBinding('class.k-input-separator'),
        tslib_1.__metadata("design:type", Boolean)
    ], InputSeparatorComponent.prototype, "hostClass", void 0);
InputSeparatorComponent.ngComponentDef = ɵngcc0.ɵɵdefineComponent({ type: InputSeparatorComponent, selectors: [["kendo-input-separator"], ["kendo-textbox-separator"]], factory: function InputSeparatorComponent_Factory(t) { return new (t || InputSeparatorComponent)(); }, hostBindings: function InputSeparatorComponent_HostBindings(rf, ctx, elIndex) { if (rf & 1) {
        ɵngcc0.ɵɵallocHostVars(1);
        ɵngcc0.ɵɵstyling();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("k-input-separator", ctx.hostClass);
        ɵngcc0.ɵɵstylingApply();
    } }, consts: 0, vars: 0, template: function InputSeparatorComponent_Template(rf, ctx) { }, encapsulation: 2 });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(InputSeparatorComponent, [{
        type: Component,
        args: [{
                selector: 'kendo-input-separator, kendo-textbox-separator',
                template: ""
            }]
    }], function () { return []; }, { hostClass: [{
            type: HostBinding,
            args: ['class.k-input-separator']
        }] });
    return InputSeparatorComponent;
}());
export { InputSeparatorComponent };
