// @progress/kendo-angular-common/dist/es/draggable/draggable.module.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { DraggableDirective } from './draggable.directive';
import { CommonModule } from '@angular/common';
/**
 * @hidden
 */
import * as ɵngcc0 from '@angular/core';
var DraggableModule = /** @class */ (function () {
    function DraggableModule() {
    }
DraggableModule.ngModuleDef = ɵngcc0.ɵɵdefineNgModule({ type: DraggableModule });
/*@__PURE__*/ ɵngcc0.ɵɵsetNgModuleScope(DraggableModule, { declarations: function () { return [DraggableDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [DraggableDirective]; } });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(DraggableModule, [{
        type: NgModule,
        args: [{
                declarations: [DraggableDirective],
                exports: [DraggableDirective],
                imports: [CommonModule]
            }]
    }], function () { return []; }, null);
DraggableModule.ngInjectorDef = ɵngcc0.ɵɵdefineInjector({ factory: function DraggableModule_Factory(t) { return new (t || DraggableModule)(); }, imports: [[CommonModule]] });
    return DraggableModule;
}());
export { DraggableModule };
