// @progress/kendo-angular-inputs/dist/es2015/checkbox/checkbox.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Directive, HostBinding } from '@angular/core';
/**
 * Represents the directive that renders the [Kendo UI CheckBox]({% slug overview_checkbox %}) input component.
 * The directive is placed on input type="checkbox" elements.
 *
 * @example
 * ```ts-no-run
 * <input type="checkbox" kendoCheckBox />
 * ```
 */
import * as ɵngcc0 from '@angular/core';
let CheckBoxDirective = class CheckBoxDirective {
    /**
     * Represents the directive that renders the [Kendo UI CheckBox]({% slug overview_checkbox %}) input component.
     * The directive is placed on input type="checkbox" elements.
     *
     * @example
     * ```ts-no-run
     * <input type="checkbox" kendoCheckBox />
     * ```
     */
    constructor() {
        this.kendoClass = true;
    }
}
CheckBoxDirective.ngDirectiveDef = ɵngcc0.ɵɵdefineDirective({ type: CheckBoxDirective, selectors: [["input", "kendoCheckBox", ""]], factory: function CheckBoxDirective_Factory(t) { return new (t || CheckBoxDirective)(); }, hostBindings: function CheckBoxDirective_HostBindings(rf, ctx, elIndex) { if (rf & 1) {
        ɵngcc0.ɵɵallocHostVars(1);
        ɵngcc0.ɵɵstyling();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("k-checkbox", ctx.kendoClass);
        ɵngcc0.ɵɵstylingApply();
    } } });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(CheckBoxDirective, [{
        type: Directive,
        args: [{
                selector: 'input[kendoCheckBox]'
            }]
    }], function () { return []; }, { constructor: [], kendoClass: [{
            type: HostBinding,
            args: ['class.k-checkbox']
        }] });;
tslib_1.__decorate([
    HostBinding('class.k-checkbox'),
    tslib_1.__metadata("design:type", Boolean)
], CheckBoxDirective.prototype, "kendoClass", void 0);
export { CheckBoxDirective };
