// @progress/kendo-angular-intl/dist/npm/intl.module.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var date_pipe_1 = require("./date.pipe");
var number_pipe_1 = require("./number.pipe");
var ɵngcc0 = require('@angular/core');
var ɵngcc1 = require('./date.pipe');
var ɵngcc2 = require('./number.pipe');
var pipes = [
    date_pipe_1.DatePipe,
    number_pipe_1.NumberPipe
];
/**
 * Represents the [NgModule]({{ site.data.urls.angular['ngmoduleapi'] }}docs/ts/latest/guide/ngmodule.html)
 * definition for the Intl services.
 */
var IntlModule = /** @class */ (function () {
    function IntlModule() {
    }
IntlModule.ngModuleDef = ɵngcc0.ɵɵdefineNgModule({ type: IntlModule });
/*@__PURE__*/ ɵngcc0.ɵɵsetNgModuleScope(IntlModule, { declarations: [ɵngcc1.DatePipe, ɵngcc2.NumberPipe], exports: [ɵngcc1.DatePipe, ɵngcc2.NumberPipe] });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(IntlModule, [{
        type: NgModule,
        args: [{
                declarations: [pipes],
                exports: [pipes]
            }]
    }], function () { return []; }, null);
IntlModule.ngInjectorDef = ɵngcc0.ɵɵdefineInjector({ factory: function IntlModule_Factory(t) { return new (t || IntlModule)(); } });
    return IntlModule;
}());
exports.IntlModule = IntlModule;
