// @progress/kendo-angular-inputs/dist/es/shared.module.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { NgModule } from "@angular/core";
import { InputSeparatorComponent } from './shared/input-separator.component';
import { TextAreaDirective } from "./shared/textarea.directive";
/**
 * @hidden
 */
import * as ɵngcc0 from '@angular/core';
export var SHARED_DIRECTIVES = [
    InputSeparatorComponent,
    TextAreaDirective
];
/**
 * @hidden
 */
var SharedModule = /** @class */ (function () {
    function SharedModule() {
    }
SharedModule.ngModuleDef = ɵngcc0.ɵɵdefineNgModule({ type: SharedModule });
/*@__PURE__*/ ɵngcc0.ɵɵsetNgModuleScope(SharedModule, { declarations: [InputSeparatorComponent,
        TextAreaDirective], exports: [InputSeparatorComponent,
        TextAreaDirective] });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(SharedModule, [{
        type: NgModule,
        args: [{
                declarations: [SHARED_DIRECTIVES],
                exports: [SHARED_DIRECTIVES]
            }]
    }], function () { return []; }, null);
SharedModule.ngInjectorDef = ɵngcc0.ɵɵdefineInjector({ factory: function SharedModule_Factory(t) { return new (t || SharedModule)(); } });
    return SharedModule;
}());
export { SharedModule };
