// @progress/kendo-angular-common/dist/es2015/resize-sensor/resize-sensor.module.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { ResizeSensorComponent } from './resize-sensor.component';
import { ResizeBatchService } from './resize-batch.service';
import * as ɵngcc0 from '@angular/core';
const COMPONENT_DIRECTIVES = [ResizeSensorComponent];
/**
 * Resize Sensor module
 */
let ResizeSensorModule = class ResizeSensorModule {
}
ResizeSensorModule.ngModuleDef = ɵngcc0.ɵɵdefineNgModule({ type: ResizeSensorModule });
/*@__PURE__*/ ɵngcc0.ɵɵsetNgModuleScope(ResizeSensorModule, { declarations: [ResizeSensorComponent], exports: [ResizeSensorComponent] });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(ResizeSensorModule, [{
        type: NgModule,
        args: [{
                declarations: [COMPONENT_DIRECTIVES],
                exports: [COMPONENT_DIRECTIVES],
                providers: [ResizeBatchService]
            }]
    }], null, null);
ResizeSensorModule.ngInjectorDef = ɵngcc0.ɵɵdefineInjector({ factory: function ResizeSensorModule_Factory(t) { return new (t || ResizeSensorModule)(); }, providers: [ResizeBatchService] });;
export { ResizeSensorModule };
