// @progress/kendo-angular-inputs/dist/es/radiobutton/radiobutton.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Directive, HostBinding } from '@angular/core';
/**
 * Represents the directive that renders the [Kendo UI RadioButton]({% slug overview_checkbox %}) input component.
 * The directive is placed on input type="radio" elements.
 *
 * @example
 * ```ts-no-run
 * <input type="radio" kendoRadioButton />
 * ```
 */
import * as ɵngcc0 from '@angular/core';
var RadioButtonDirective = /** @class */ (function () {
    function RadioButtonDirective() {
        this.kendoClass = true;
    }
    tslib_1.__decorate([
        HostBinding('class.k-radio'),
        tslib_1.__metadata("design:type", Boolean)
    ], RadioButtonDirective.prototype, "kendoClass", void 0);
RadioButtonDirective.ngDirectiveDef = ɵngcc0.ɵɵdefineDirective({ type: RadioButtonDirective, selectors: [["input", "kendoRadioButton", ""]], factory: function RadioButtonDirective_Factory(t) { return new (t || RadioButtonDirective)(); }, hostBindings: function RadioButtonDirective_HostBindings(rf, ctx, elIndex) { if (rf & 1) {
        ɵngcc0.ɵɵallocHostVars(1);
        ɵngcc0.ɵɵstyling();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("k-radio", ctx.kendoClass);
        ɵngcc0.ɵɵstylingApply();
    } } });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(RadioButtonDirective, [{
        type: Directive,
        args: [{
                selector: 'input[kendoRadioButton]'
            }]
    }], function () { return []; }, { kendoClass: [{
            type: HostBinding,
            args: ['class.k-radio']
        }] });
    return RadioButtonDirective;
}());
export { RadioButtonDirective };
