// @progress/kendo-angular-inputs/dist/es2015/colorpicker.module.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PopupModule } from '@progress/kendo-angular-popup';
import { DraggableModule } from '@progress/kendo-angular-common';
import { SliderModule } from './slider.module';
import { NumericTextBoxModule } from './numerictextbox.module';
import { ColorPickerComponent } from './colorpicker/colorpicker.component';
import { ColorPaletteComponent } from './colorpicker/color-palette.component';
import { ColorGradientComponent } from './colorpicker/color-gradient.component';
import { ColorInputComponent } from './colorpicker/color-input.component';
import { ColorPickerCustomMessagesComponent } from './colorpicker/localization/custom-messages.component';
import { LocalizedColorPickerMessagesDirective } from './colorpicker/localization/localized-colorpicker-messages.directive';
import { FocusOnDomReadyDirective } from './colorpicker/focus-on-dom-ready.directive';
import { ContrastValidationComponent } from './colorpicker/contrast-validation.component';
import { ContrastComponent } from './colorpicker/contrast.component';
import * as ɵngcc0 from '@angular/core';
const PUBLIC_DIRECTIVES = [
    ColorPickerComponent,
    ColorPaletteComponent,
    ColorGradientComponent,
    LocalizedColorPickerMessagesDirective,
    ColorPickerCustomMessagesComponent
];
const INTERNAL_DIRECTIVES = [
    ColorInputComponent,
    FocusOnDomReadyDirective,
    ContrastComponent,
    ContrastValidationComponent
];
/**
 * Represents the [NgModule]({{ site.data.urls.angular['ngmoduleapi'] }})
 * definition for the ColorPicker.
 */
let ColorPickerModule = class ColorPickerModule {
}
ColorPickerModule.ngModuleDef = ɵngcc0.ɵɵdefineNgModule({ type: ColorPickerModule });
/*@__PURE__*/ ɵngcc0.ɵɵsetNgModuleScope(ColorPickerModule, { declarations: function () { return [ColorPickerComponent,
        ColorPaletteComponent,
        ColorGradientComponent,
        LocalizedColorPickerMessagesDirective,
        ColorPickerCustomMessagesComponent,
        ColorInputComponent,
        FocusOnDomReadyDirective,
        ContrastComponent,
        ContrastValidationComponent]; }, imports: function () { return [SliderModule,
        NumericTextBoxModule,
        CommonModule,
        PopupModule,
        DraggableModule]; }, exports: function () { return [ColorPickerComponent,
        ColorPaletteComponent,
        ColorGradientComponent,
        LocalizedColorPickerMessagesDirective,
        ColorPickerCustomMessagesComponent]; } });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(ColorPickerModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    PUBLIC_DIRECTIVES,
                    INTERNAL_DIRECTIVES
                ],
                exports: [PUBLIC_DIRECTIVES],
                imports: [
                    SliderModule,
                    NumericTextBoxModule,
                    CommonModule,
                    PopupModule,
                    DraggableModule
                ]
            }]
    }], null, null);
ColorPickerModule.ngInjectorDef = ɵngcc0.ɵɵdefineInjector({ factory: function ColorPickerModule_Factory(t) { return new (t || ColorPickerModule)(); }, imports: [[
            SliderModule,
            NumericTextBoxModule,
            CommonModule,
            PopupModule,
            DraggableModule
        ]] });;
export { ColorPickerModule };
