// @progress/kendo-angular-inputs/dist/es2015/textarea.module.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { TextAreaComponent } from "./textarea/textarea.component";
import { EventsModule } from "@progress/kendo-angular-common";
import { TextAreaSuffixComponent } from './textarea/textarea-suffix.component';
import { SharedModule, SHARED_DIRECTIVES } from './shared.module';
/**
 * Represents the [NgModule]({{ site.data.urls.angular['ngmoduleapi'] }})
 * definition for the TextArea component.
 *
 * @example
 *
 * ```ts-no-run
 * // Import the TextArea module
 * import { TextAreaModule } from '@progress/kendo-angular-inputs';
 *
 * // The browser platform with a compiler
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 *
 * import { NgModule } from '@angular/core';
 *
 * // Import the app component
 * import { AppComponent } from './app.component';
 *
 * // Define the app module
 * _@NgModule({
 *     declarations: [AppComponent], // declare app component
 *     imports:      [BrowserModule, TextAreaModule], // import TextArea module
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * // Compile and launch the module
 * platformBrowserDynamic().bootstrapModule(AppModule);
 *
 * ```
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './shared/input-separator.component';
import * as ɵngcc2 from './shared/textarea.directive';
let TextAreaModule = class TextAreaModule {
}
TextAreaModule.ngModuleDef = ɵngcc0.ɵɵdefineNgModule({ type: TextAreaModule });
/*@__PURE__*/ ɵngcc0.ɵɵsetNgModuleScope(TextAreaModule, { declarations: function () { return [TextAreaComponent,
        TextAreaSuffixComponent]; }, imports: function () { return [CommonModule, EventsModule, SharedModule]; }, exports: function () { return [TextAreaComponent,
        EventsModule,
        TextAreaSuffixComponent, ɵngcc1.InputSeparatorComponent, ɵngcc2.TextAreaDirective]; } });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(TextAreaModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    TextAreaComponent,
                    TextAreaSuffixComponent
                ],
                exports: [
                    TextAreaComponent,
                    EventsModule,
                    TextAreaSuffixComponent,
                    SHARED_DIRECTIVES
                ],
                imports: [CommonModule, EventsModule, SharedModule]
            }]
    }], null, null);
TextAreaModule.ngInjectorDef = ɵngcc0.ɵɵdefineInjector({ factory: function TextAreaModule_Factory(t) { return new (t || TextAreaModule)(); }, imports: [[CommonModule, EventsModule, SharedModule]] });;
export { TextAreaModule };
