// @progress/kendo-angular-inputs/dist/es2015/textarea/textarea-suffix.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Component, HostBinding } from '@angular/core';
/**
 * Specifies the adornments in the suffix container ([see example]({% slug textarea_adornments %}).
 *  @example
 * ```ts-no-run
 * _@Component({
 * selector: 'my-app',
 * template: `
 *  <kendo-textarea>
 *    <kendo-textarea-suffix>
 *        <button kendoButton look="clear" icon="image"></button>
 *    </kendo-textarea-suffix>
 *  </kendo-textarea>
 * `
 * })
 * class AppComponent {}
 * ```
 */
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
let TextAreaSuffixComponent = class TextAreaSuffixComponent {
    /**
     * Specifies the adornments in the suffix container ([see example]({% slug textarea_adornments %}).
     *  @example
     * ```ts-no-run
     * _@Component({
     * selector: 'my-app',
     * template: `
     *  <kendo-textarea>
     *    <kendo-textarea-suffix>
     *        <button kendoButton look="clear" icon="image"></button>
     *    </kendo-textarea-suffix>
     *  </kendo-textarea>
     * `
     * })
     * class AppComponent {}
     * ```
     */
    constructor() {
        this.hostClass = true;
    }
}
TextAreaSuffixComponent.ngComponentDef = ɵngcc0.ɵɵdefineComponent({ type: TextAreaSuffixComponent, selectors: [["kendo-textarea-suffix"]], factory: function TextAreaSuffixComponent_Factory(t) { return new (t || TextAreaSuffixComponent)(); }, hostBindings: function TextAreaSuffixComponent_HostBindings(rf, ctx, elIndex) { if (rf & 1) {
        ɵngcc0.ɵɵallocHostVars(1);
        ɵngcc0.ɵɵstyling();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("k-textarea-suffix", ctx.hostClass);
        ɵngcc0.ɵɵstylingApply();
    } }, exportAs: ["kendoTextAreaSuffix"], ngContentSelectors: _c0, consts: 1, vars: 0, template: function TextAreaSuffixComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(TextAreaSuffixComponent, [{
        type: Component,
        args: [{
                exportAs: 'kendoTextAreaSuffix',
                selector: 'kendo-textarea-suffix',
                template: `<ng-content></ng-content>`
            }]
    }], function () { return []; }, { constructor: [], hostClass: [{
            type: HostBinding,
            args: ['class.k-textarea-suffix']
        }] });;
tslib_1.__decorate([
    HostBinding('class.k-textarea-suffix'),
    tslib_1.__metadata("design:type", Boolean)
], TextAreaSuffixComponent.prototype, "hostClass", void 0);
export { TextAreaSuffixComponent };
