// @progress/kendo-angular-inputs/dist/es/colorpicker/focus-on-dom-ready.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Directive, ElementRef, NgZone } from '@angular/core';
/**
 * @hidden
 */
import * as ɵngcc0 from '@angular/core';
var FocusOnDomReadyDirective = /** @class */ (function () {
    function FocusOnDomReadyDirective(host, ngZone) {
        this.host = host;
        this.ngZone = ngZone;
    }
    FocusOnDomReadyDirective.prototype.ngAfterContentInit = function () {
        this.focusOnNextTick();
    };
    FocusOnDomReadyDirective.prototype.focusOnNextTick = function () {
        var _this = this;
        this.ngZone.runOutsideAngular(function () { return setTimeout(function () { return _this.host.nativeElement.focus(); }); });
    };
    FocusOnDomReadyDirective = tslib_1.__decorate([
        tslib_1.__metadata("design:paramtypes", [ElementRef,
            NgZone])
    ], FocusOnDomReadyDirective);
FocusOnDomReadyDirective.ngDirectiveDef = ɵngcc0.ɵɵdefineDirective({ type: FocusOnDomReadyDirective, selectors: [["", "kendoFocusOnDomReady", ""]], factory: function FocusOnDomReadyDirective_Factory(t) { return new (t || FocusOnDomReadyDirective)(ɵngcc0.ɵɵdirectiveInject(ElementRef), ɵngcc0.ɵɵdirectiveInject(NgZone)); } });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(FocusOnDomReadyDirective, [{
        type: Directive,
        args: [{
                selector: '[kendoFocusOnDomReady]'
            }]
    }], function () { return [{ type: ElementRef }, { type: NgZone }]; }, { host: [], ngZone: [], ngAfterContentInit: [], focusOnNextTick: [] });
    return FocusOnDomReadyDirective;
}());
export { FocusOnDomReadyDirective };
