// @progress/kendo-angular-inputs/dist/es/colorpicker/localization/custom-messages.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { ColorPickerMessages } from './messages';
/**
 * Custom component messages override default component messages.
 */
import * as ɵngcc0 from '@angular/core';
var ColorPickerCustomMessagesComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ColorPickerCustomMessagesComponent, _super);
    function ColorPickerCustomMessagesComponent(service) {
        var _this = _super.call(this) || this;
        _this.service = service;
        return _this;
    }
    ColorPickerCustomMessagesComponent_1 = ColorPickerCustomMessagesComponent;
    Object.defineProperty(ColorPickerCustomMessagesComponent.prototype, "override", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    var ColorPickerCustomMessagesComponent_1;
    ColorPickerCustomMessagesComponent = ColorPickerCustomMessagesComponent_1 = tslib_1.__decorate([
        tslib_1.__metadata("design:paramtypes", [LocalizationService])
    ], ColorPickerCustomMessagesComponent);
ColorPickerCustomMessagesComponent.ngComponentDef = ɵngcc0.ɵɵdefineComponent({ type: ColorPickerCustomMessagesComponent, selectors: [["kendo-colorpicker-messages"], ["kendo-colorgradient-messages"], ["kendo-colorpalette-messages"]], factory: function ColorPickerCustomMessagesComponent_Factory(t) { return new (t || ColorPickerCustomMessagesComponent)(ɵngcc0.ɵɵdirectiveInject(LocalizationService)); }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: ColorPickerMessages,
                useExisting: forwardRef(function () { return ColorPickerCustomMessagesComponent_1; }) // tslint:disable-line:no-forward-ref
            }
        ]), ɵngcc0.ɵɵInheritDefinitionFeature], consts: 0, vars: 0, template: function ColorPickerCustomMessagesComponent_Template(rf, ctx) { }, encapsulation: 2 });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(ColorPickerCustomMessagesComponent, [{
        type: Component,
        args: [{
                providers: [
                    {
                        provide: ColorPickerMessages,
                        useExisting: forwardRef(function () { return ColorPickerCustomMessagesComponent_1; }) // tslint:disable-line:no-forward-ref
                    }
                ],
                selector: 'kendo-colorpicker-messages, kendo-colorgradient-messages, kendo-colorpalette-messages',
                template: ""
            }]
    }], function () { return [{ type: LocalizationService }]; }, { override: [] });
    return ColorPickerCustomMessagesComponent;
}(ColorPickerMessages));
export { ColorPickerCustomMessagesComponent };
