// @progress/kendo-angular-inputs/dist/es/colorpicker/localization/localized-colorpicker-messages.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Directive, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { ColorPickerMessages } from './messages';
/**
 * @hidden
 */
import * as ɵngcc0 from '@angular/core';
var LocalizedColorPickerMessagesDirective = /** @class */ (function (_super) {
    tslib_1.__extends(LocalizedColorPickerMessagesDirective, _super);
    function LocalizedColorPickerMessagesDirective(service) {
        var _this = _super.call(this) || this;
        _this.service = service;
        return _this;
    }
    LocalizedColorPickerMessagesDirective_1 = LocalizedColorPickerMessagesDirective;
    var LocalizedColorPickerMessagesDirective_1;
    LocalizedColorPickerMessagesDirective = LocalizedColorPickerMessagesDirective_1 = tslib_1.__decorate([
        tslib_1.__metadata("design:paramtypes", [LocalizationService])
    ], LocalizedColorPickerMessagesDirective);
LocalizedColorPickerMessagesDirective.ngDirectiveDef = ɵngcc0.ɵɵdefineDirective({ type: LocalizedColorPickerMessagesDirective, selectors: [["", "kendoColorPickerLocalizedMessages", ""], ["", "kendoColorGradientLocalizedMessages", ""], ["", "kendoColorPaletteLocalizedMessages", ""]], factory: function LocalizedColorPickerMessagesDirective_Factory(t) { return new (t || LocalizedColorPickerMessagesDirective)(ɵngcc0.ɵɵdirectiveInject(LocalizationService)); }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: ColorPickerMessages,
                useExisting: forwardRef(function () { return LocalizedColorPickerMessagesDirective_1; }) // tslint:disable-line:no-forward-ref
            }
        ]), ɵngcc0.ɵɵInheritDefinitionFeature] });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(LocalizedColorPickerMessagesDirective, [{
        type: Directive,
        args: [{
                providers: [
                    {
                        provide: ColorPickerMessages,
                        useExisting: forwardRef(function () { return LocalizedColorPickerMessagesDirective_1; }) // tslint:disable-line:no-forward-ref
                    }
                ],
                selector: '[kendoColorPickerLocalizedMessages], [kendoColorGradientLocalizedMessages], [kendoColorPaletteLocalizedMessages]'
            }]
    }], function () { return [{ type: LocalizationService }]; }, null);
    return LocalizedColorPickerMessagesDirective;
}(ColorPickerMessages));
export { LocalizedColorPickerMessagesDirective };
