// @progress/kendo-angular-inputs/dist/es2015/colorpicker/contrast.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { Component, Input } from '@angular/core';
import { getContrastFromTwoRGBAs, getRGBA } from './utils';
import { AA_RATIO, AAA_RATIO } from './constants';
/**
 * @hidden
 */
import * as ɵngcc0 from '@angular/core';

const _c0 = ["kendoContrastTool", ""];
const _c1 = [1, "k-contrast-ratio"];
const _c2 = [1, "k-contrast-ratio-text"];
const _c3 = ["class", "k-contrast-validation k-text-success", 4, "ngIf"];
const _c4 = ["class", "k-contrast-validation k-text-error", 4, "ngIf"];
const _c5 = ["kendoContrastValidation", "", "type", "AA", 3, "ratio", "pass"];
const _c6 = ["kendoContrastValidation", "", "type", "AAA", 3, "ratio", "pass"];
const _c7 = [1, "k-contrast-validation", "k-text-success"];
const _c8 = [1, "k-icon", "k-i-check"];
const _c9 = ["class", "k-icon k-i-check", 4, "ngIf"];
function ContrastComponent_span_3_span_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", _c8);
} }
function ContrastComponent_span_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", _c7);
    ɵngcc0.ɵɵelement(1, "span", _c8);
    ɵngcc0.ɵɵtemplate(2, ContrastComponent_span_3_span_2_Template, 1, 0, "span", _c9);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r263 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵselect(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r263.satisfiesAAACondition);
} }
const _c10 = [1, "k-contrast-validation", "k-text-error"];
const _c11 = [1, "k-icon", "k-i-close"];
function ContrastComponent_span_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", _c10);
    ɵngcc0.ɵɵelement(1, "span", _c11);
    ɵngcc0.ɵɵelementEnd();
} }
let ContrastComponent = class ContrastComponent {
    constructor(localization) {
        this.localization = localization;
        this.aaRatio = AA_RATIO;
        this.aaaRatio = AAA_RATIO;
    }
    get formatedRatio() {
        return this.contrastRatio.toFixed(2);
    }
    get contrastRatioMessage() {
        return this.localization.get('contrastRatio');
    }
    get satisfiesAACondition() {
        return this.contrastRatio >= this.aaRatio;
    }
    get satisfiesAAACondition() {
        return this.contrastRatio >= this.aaaRatio;
    }
    get contrastRatio() {
        let contrast = getContrastFromTwoRGBAs(getRGBA(this.value), getRGBA(this.ratio));
        return contrast;
    }
}
ContrastComponent.ngComponentDef = ɵngcc0.ɵɵdefineComponent({ type: ContrastComponent, selectors: [["", "kendoContrastTool", ""]], factory: function ContrastComponent_Factory(t) { return new (t || ContrastComponent)(ɵngcc0.ɵɵdirectiveInject(LocalizationService)); }, inputs: { value: "value", ratio: "ratio" }, attrs: _c0, consts: 7, vars: 8, template: function ContrastComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", _c1);
        ɵngcc0.ɵɵelementStart(1, "span", _c2);
        ɵngcc0.ɵɵtext(2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(3, ContrastComponent_span_3_Template, 3, 1, "span", _c3);
        ɵngcc0.ɵɵtemplate(4, ContrastComponent_span_4_Template, 2, 0, "span", _c4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(5, "div", _c5);
        ɵngcc0.ɵɵelement(6, "div", _c6);
    } if (rf & 2) {
        ɵngcc0.ɵɵselect(2);
        ɵngcc0.ɵɵtextInterpolate2("", ctx.contrastRatioMessage, ": ", ctx.formatedRatio, "");
        ɵngcc0.ɵɵselect(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.satisfiesAACondition);
        ɵngcc0.ɵɵselect(4);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.satisfiesAACondition);
        ɵngcc0.ɵɵselect(5);
        ɵngcc0.ɵɵproperty("ratio", ctx.aaaRatio)("pass", ctx.satisfiesAACondition);
        ɵngcc0.ɵɵselect(6);
        ɵngcc0.ɵɵproperty("ratio", ctx.aaaRatio)("pass", ctx.satisfiesAAACondition);
    } }, encapsulation: 2 });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(ContrastComponent, [{
        type: Component,
        args: [{
                selector: '[kendoContrastTool]',
                template: `
        <div class="k-contrast-ratio">
            <span class="k-contrast-ratio-text">{{contrastRatioMessage}}: {{formatedRatio}}</span>
            <span class="k-contrast-validation k-text-success" *ngIf="satisfiesAACondition">
                <span class="k-icon k-i-check"></span>
                <span class="k-icon k-i-check" *ngIf="satisfiesAAACondition"></span>
            </span>
            <span class="k-contrast-validation k-text-error" *ngIf="!satisfiesAACondition">
                <span class="k-icon k-i-close"></span>
            </span>
        </div>
        <div kendoContrastValidation
            type="AA"
            [ratio]="aaaRatio"
            [pass]="satisfiesAACondition">
        </div>
        <div kendoContrastValidation
            type="AAA"
            [ratio]="aaaRatio"
            [pass]="satisfiesAAACondition">
        </div>
    `
            }]
    }], function () { return [{ type: LocalizationService }]; }, { constructor: [], localization: [], aaRatio: [], aaaRatio: [], formatedRatio: [], contrastRatioMessage: [], satisfiesAACondition: [], satisfiesAAACondition: [], contrastRatio: [], value: [{
            type: Input
        }], ratio: [{
            type: Input
        }] });;
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], ContrastComponent.prototype, "value", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], ContrastComponent.prototype, "ratio", void 0);
ContrastComponent = tslib_1.__decorate([
    tslib_1.__metadata("design:paramtypes", [LocalizationService])
], ContrastComponent);
export { ContrastComponent };
