// @progress/kendo-angular-inputs/dist/es2015/colorpicker/focus-on-dom-ready.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Directive, ElementRef, NgZone } from '@angular/core';
/**
 * @hidden
 */
import * as ɵngcc0 from '@angular/core';
let FocusOnDomReadyDirective = class FocusOnDomReadyDirective {
    constructor(host, ngZone) {
        this.host = host;
        this.ngZone = ngZone;
    }
    ngAfterContentInit() {
        this.focusOnNextTick();
    }
    focusOnNextTick() {
        this.ngZone.runOutsideAngular(() => setTimeout(() => this.host.nativeElement.focus()));
    }
}
FocusOnDomReadyDirective.ngDirectiveDef = ɵngcc0.ɵɵdefineDirective({ type: FocusOnDomReadyDirective, selectors: [["", "kendoFocusOnDomReady", ""]], factory: function FocusOnDomReadyDirective_Factory(t) { return new (t || FocusOnDomReadyDirective)(ɵngcc0.ɵɵdirectiveInject(ElementRef), ɵngcc0.ɵɵdirectiveInject(NgZone)); } });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(FocusOnDomReadyDirective, [{
        type: Directive,
        args: [{
                selector: '[kendoFocusOnDomReady]'
            }]
    }], function () { return [{ type: ElementRef }, { type: NgZone }]; }, { constructor: [], host: [], ngZone: [], ngAfterContentInit: [], focusOnNextTick: [] });;
FocusOnDomReadyDirective = tslib_1.__decorate([
    tslib_1.__metadata("design:paramtypes", [ElementRef,
        NgZone])
], FocusOnDomReadyDirective);
export { FocusOnDomReadyDirective };
