// @progress/kendo-angular-inputs/dist/es/numerictextbox/localization/custom-messages.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { NumericTextBoxMessages } from './messages';
/**
 * Custom component messages override default component messages.
 */
import * as ɵngcc0 from '@angular/core';
var NumericTextBoxCustomMessagesComponent = /** @class */ (function (_super) {
    tslib_1.__extends(NumericTextBoxCustomMessagesComponent, _super);
    function NumericTextBoxCustomMessagesComponent(service) {
        var _this = _super.call(this) || this;
        _this.service = service;
        return _this;
    }
    NumericTextBoxCustomMessagesComponent_1 = NumericTextBoxCustomMessagesComponent;
    Object.defineProperty(NumericTextBoxCustomMessagesComponent.prototype, "override", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    var NumericTextBoxCustomMessagesComponent_1;
    NumericTextBoxCustomMessagesComponent = NumericTextBoxCustomMessagesComponent_1 = tslib_1.__decorate([
        tslib_1.__metadata("design:paramtypes", [LocalizationService])
    ], NumericTextBoxCustomMessagesComponent);
NumericTextBoxCustomMessagesComponent.ngComponentDef = ɵngcc0.ɵɵdefineComponent({ type: NumericTextBoxCustomMessagesComponent, selectors: [["kendo-numerictextbox-messages"]], factory: function NumericTextBoxCustomMessagesComponent_Factory(t) { return new (t || NumericTextBoxCustomMessagesComponent)(ɵngcc0.ɵɵdirectiveInject(LocalizationService)); }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: NumericTextBoxMessages,
                useExisting: forwardRef(function () { return NumericTextBoxCustomMessagesComponent_1; }) // tslint:disable-line:no-forward-ref
            }
        ]), ɵngcc0.ɵɵInheritDefinitionFeature], consts: 0, vars: 0, template: function NumericTextBoxCustomMessagesComponent_Template(rf, ctx) { }, encapsulation: 2 });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(NumericTextBoxCustomMessagesComponent, [{
        type: Component,
        args: [{
                providers: [
                    {
                        provide: NumericTextBoxMessages,
                        useExisting: forwardRef(function () { return NumericTextBoxCustomMessagesComponent_1; }) // tslint:disable-line:no-forward-ref
                    }
                ],
                selector: 'kendo-numerictextbox-messages',
                template: ""
            }]
    }], function () { return [{ type: LocalizationService }]; }, { override: [] });
    return NumericTextBoxCustomMessagesComponent;
}(NumericTextBoxMessages));
export { NumericTextBoxCustomMessagesComponent };
