// @progress/kendo-angular-inputs/dist/es/numerictextbox/localization/localized-numerictextbox-messages.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { Directive, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { NumericTextBoxMessages } from './messages';
/**
 * @hidden
 */
import * as ɵngcc0 from '@angular/core';
var LocalizedNumericTextBoxMessagesDirective = /** @class */ (function (_super) {
    tslib_1.__extends(LocalizedNumericTextBoxMessagesDirective, _super);
    function LocalizedNumericTextBoxMessagesDirective(service) {
        var _this = _super.call(this) || this;
        _this.service = service;
        return _this;
    }
    LocalizedNumericTextBoxMessagesDirective_1 = LocalizedNumericTextBoxMessagesDirective;
    var LocalizedNumericTextBoxMessagesDirective_1;
    LocalizedNumericTextBoxMessagesDirective = LocalizedNumericTextBoxMessagesDirective_1 = tslib_1.__decorate([
        tslib_1.__metadata("design:paramtypes", [LocalizationService])
    ], LocalizedNumericTextBoxMessagesDirective);
LocalizedNumericTextBoxMessagesDirective.ngDirectiveDef = ɵngcc0.ɵɵdefineDirective({ type: LocalizedNumericTextBoxMessagesDirective, selectors: [["", "kendoNumericTextBoxLocalizedMessages", ""]], factory: function LocalizedNumericTextBoxMessagesDirective_Factory(t) { return new (t || LocalizedNumericTextBoxMessagesDirective)(ɵngcc0.ɵɵdirectiveInject(LocalizationService)); }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: NumericTextBoxMessages,
                useExisting: forwardRef(function () { return LocalizedNumericTextBoxMessagesDirective_1; }) // tslint:disable-line:no-forward-ref
            }
        ]), ɵngcc0.ɵɵInheritDefinitionFeature] });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(LocalizedNumericTextBoxMessagesDirective, [{
        type: Directive,
        args: [{
                providers: [
                    {
                        provide: NumericTextBoxMessages,
                        useExisting: forwardRef(function () { return LocalizedNumericTextBoxMessagesDirective_1; }) // tslint:disable-line:no-forward-ref
                    }
                ],
                selector: '[kendoNumericTextBoxLocalizedMessages]'
            }]
    }], function () { return [{ type: LocalizationService }]; }, null);
    return LocalizedNumericTextBoxMessagesDirective;
}(NumericTextBoxMessages));
export { LocalizedNumericTextBoxMessagesDirective };
