// @progress/kendo-angular-inputs/dist/es/sliders-common/sliders-common.module.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { SliderTicksComponent } from './slider-ticks.component';
import { LabelTemplateDirective } from './label-template.directive';
import { DraggableModule, EventsModule, ResizeSensorModule } from '@progress/kendo-angular-common';
/**
 * @hidden
 */
import * as ɵngcc0 from '@angular/core';
var SlidersCommonModule = /** @class */ (function () {
    function SlidersCommonModule() {
    }
SlidersCommonModule.ngModuleDef = ɵngcc0.ɵɵdefineNgModule({ type: SlidersCommonModule });
/*@__PURE__*/ ɵngcc0.ɵɵsetNgModuleScope(SlidersCommonModule, { declarations: function () { return [SliderTicksComponent,
        LabelTemplateDirective]; }, imports: function () { return [CommonModule, DraggableModule, EventsModule, ResizeSensorModule]; }, exports: function () { return [LabelTemplateDirective,
        SliderTicksComponent,
        DraggableModule,
        EventsModule,
        ResizeSensorModule]; } });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(SlidersCommonModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    SliderTicksComponent,
                    LabelTemplateDirective
                ],
                exports: [
                    LabelTemplateDirective,
                    SliderTicksComponent,
                    DraggableModule,
                    EventsModule,
                    ResizeSensorModule
                ],
                imports: [CommonModule, DraggableModule, EventsModule, ResizeSensorModule]
            }]
    }], function () { return []; }, null);
SlidersCommonModule.ngInjectorDef = ɵngcc0.ɵɵdefineInjector({ factory: function SlidersCommonModule_Factory(t) { return new (t || SlidersCommonModule)(); }, imports: [[CommonModule, DraggableModule, EventsModule, ResizeSensorModule],
        DraggableModule,
        EventsModule,
        ResizeSensorModule] });
    return SlidersCommonModule;
}());
export { SlidersCommonModule };
